autobahn/asyncio/component.py

```diff
@@ -312,101 +315,118 @@ def start(self, loop=None):
+            valid_transports = [tp for tp in self._transports if tp.can_reconnect()]
+            connectable_transports['count'] =  len(valid_transports)
+            connectable_transports['gen'] = itertools.cycle(valid_transports)
+
+        set_connectable_transports()
 
         # issue our first event, then start the reconnect loop
         f0 = self.fire('start', loop, self)
 
-        # this is a 1-element list so we can set it from closures in
-        # this function
-        reconnect = [True]
-
         def one_reconnect_loop(_):
             self.log.debug('Entering re-connect loop')
-            if not reconnect[0]:
-                return
-
-            # cycle through all transports forever ..
-            transport = next(transport_gen)
-
-            # only actually try to connect using the transport,
-            # if the transport hasn't reached max. connect count
-            if transport.can_reconnect():
-                delay = transport.next_delay()
-                self.log.debug(
-                    'trying transport {transport_idx} using connect delay {transport_delay}',
-                    transport_idx=transport.idx,
-                    transport_delay=delay,
-                )
 
-                delay_f = asyncio.ensure_future(txaio.sleep(delay))
+            if not connectable_transports['count']:
+                err_msg = u"Component failed: Exhausted all transport connect attempts"
+                self.log.info(err_msg)
+                try:
+                    raise RuntimeError(err_msg)
+                except RuntimeError as e:
+                    txaio.reject(done_f, e)
+                    return
+
+            def actual_connect(_):
+
+                def handle_connect_error(fail):
+                    unrecoverable_error = False
+
+                    if isinstance(fail.value, asyncio.CancelledError):
+                        unrecoverable_error = True
+
+                    self.log.debug(u'component failed: {error}', error=txaio.failure_message(fail))
+                    self.log.debug(u'{tb}', tb=txaio.failure_format_traceback(fail))
+                    # If this is a "fatal error" that will never work,
+                    # we bail out now
+                    if isinstance(fail.value, ApplicationError):
+                        if fail.value.error in [u'wamp.error.no_such_realm']:
+                            unrecoverable_error = True
+                            self.log.error(u"Fatal error, not reconnecting")
+
+                        self.log.error(u"{msg}", msg=fail.value.error_message())
+
+                    elif isinstance(fail.value, OSError):
+                        # failed to connect entirely, like nobody
+                        # listening etc.
+                        self.log.info(u"Connection failed: {msg}", msg=txaio.failure_message(fail))
+
+                    elif _is_ssl_error(fail.value):
+                        # Quoting pyOpenSSL docs: "Whenever
```

Comment:
(Optional for this PR but ..) this code all mirrors the Twisted side -- if it could be moved to a base-class helper (or just bare function in autobahn.wamp.component) that would be ideal

Reply:
OK I'll see what can be consolidated into the base class.  Maybe most of the start method if the full mind-bending horrors of txaio can be faced in order to replace the @inlineCallbacks code in the twisted version?  Doable you think?

Reply:
If it *is* doable, that would be awesome! Even if we can move "some" of the code to a common place, that's a big win IMO.